modules/imgproc/src/shapedescr.cpp

```diff
@@ -454,6 +455,369 @@ cv::RotatedRect cv::fitEllipse( InputArray _points )
+     coeffsD[5] > (-(coeffsD[2]*(coeffsD[3]*coeffsD[3])) + coeffsD[1]*coeffsD[3]*coeffsD[4] - coeffsD[0]*(coeffsD[4]*coeffsD[4]))/
+     ((coeffsD[1]*coeffsD[1]) - 4*coeffsD[0]*coeffsD[2])) ||
+    (coeffsD[0] > 0 && coeffsD[2] > (coeffsD[1]*coeffsD[1])/(4.*coeffsD[0]) &&
+     coeffsD[5] < (-(coeffsD[2]*(coeffsD[3]*coeffsD[3])) + coeffsD[1]*coeffsD[3]*coeffsD[4] - coeffsD[0]*(coeffsD[4]*coeffsD[4]))/
+     ((coeffsD[1]*coeffsD[1]) - 4*coeffsD[0]*coeffsD[2]));
+    if (is_ellipse) {
+
+        double    u1 = pVecD[2]*pVecD[3]*pVecD[3] - pVecD[1]*pVecD[3]*pVecD[4] + pVecD[0]*pVecD[4]*pVecD[4] + pVecD[1]*pVecD[1]*coeffsD[5];
+        double    u2 = pVecD[0]*pVecD[2]*coeffsD[5];
+        double    l1 = sqrt(pVecD[1]*pVecD[1] + (pVecD[0] - pVecD[2])*(pVecD[0] - pVecD[2]));
+        double    l2 = pVecD[0] + pVecD[2];
+        double    l3 = pVecD[1]*pVecD[1] - 4.0*pVecD[0]*pVecD[2];
+        double    p1 = 2.0*pVecD[2]*pVecD[3] - pVecD[1]*pVecD[4];
+        double    p2 = 2.0*pVecD[0]*pVecD[4]-(pVecD[1]*pVecD[3]);
+
+        x0 = p1/l3 + c.x;
+        y0 = p2/l3 + c.y;
+        a = sqrt(2)*sqrt((u1 - 4.0*u2)/((l1 - l2)*l3));
+        b = sqrt(2)*sqrt(-1.0*((u1 - 4.0*u2)/((l1 + l2)*l3)));
+        if (pVecD[1] == 0) {
+            if (pVecD[0] < pVecD[2]) {
+                theta = 0;
+            } else {
+                theta = CV_PI/2.;
+            }
+        } else {
+            theta = CV_PI/2. + 0.5*std::atan2(pVecD[1], (pVecD[0] - pVecD[2]));
+        }
+
+        box.center.x = (float)x0; // +c.x;
+        box.center.y = (float)y0; // +c.y;
+        box.size.width = (float)(2.0*a);
+        box.size.height = (float)(2.0*b);
+        if( box.size.width > box.size.height )
+        {
+            float tmp;
+            CV_SWAP( box.size.width, box.size.height, tmp );
+            box.angle = (float)(90 + theta*180/CV_PI);
+        } else {
+            box.angle = (float)(fmod(theta*180/CV_PI,180.0));
+        };
+
+
+    } else {
+        box = cv::fitEllipse( points );
+
+    }    } else {
+        box = cv::fitEllipse( points );
+    }
+
+    return box;
+}
+
+cv::RotatedRect cv::fitEllipseDirect( InputArray _points )
+{
+    Mat points = _points.getMat();
+    int i, n = points.checkVector(2);
+    int depth = points.depth();
+    CV_Assert( n >= 0 && (depth == CV_32F || depth == CV_32S));
+
+    RotatedRect box;
+
+    if( n < 5 )
+        CV_Error( CV_StsBadSize, "There should be at least 5 points to fit the ellipse" );
+
+    // New fitellipse algorithm, contributed by Dr. Daniel Weiss
+    Point2f c(0,0);
+
+    bool is_float = depth == CV_32F;
+    const Point* ptsi = points.ptr<Point>();
+    const Point2f* ptsf = points.ptr<Point2f>();
+
+    AutoBuffer<double> _Ad(n*6);
+    double *Ad = _Ad;
+    Mat A( n, 6, CV_64F, Ad );
+
+    AutoBuffer<double> _DMd(6*6);
+    double *DMd = _DMd;
+    Mat DM( 6, 6, CV_64F, DMd );
+
+    AutoBuffer<double> _Md(3*3);
+    double  *Md = _Md;
+    Mat M( 3, 3, CV_64F, Md );
```

Comment:
Consider to use `Matx33d` in case of fixed dimensions.

For 6x6 matrix use this: `Matx<double, 6, 6>`

Also replace access expressions:
```.diff
-Md[1*3+0]
+M(1, 0)
```

Reply:
`AutoBuffer` and following pointer should be removed too (they are not used).